### ISSUE TITLE
OOM Error after long training

### ISSUE THREAD

**Issue by martinpopel** [tensorflow/tensor2tensor#581]
If I set `batch_size` too high, the training fails with Out Of Memory error. This is expected. The annoying point is that the failure can happen e.g. **after 3 hours of training** (15k steps). Thus it is difficult to find the highest possible batch size (for a given model, optimizer and max_length, which are other hyperparameters influencing the memory consumption I am aware of).

When I set `batch_size` *much too high*, the training fails immediately (i.e. in two minutes, after building the graph on GPU and saving the first checkpoint). However, when I set `batch_size` just *a bit too high*, the failure comes usually in the first few thousand steps, but sometimes much later (the 3 hours, 15000 steps). Moreover, it seems the number of steps before OOM failure is non-deterministic and it is not always true that higher `batch_size` fails earlier. Also it is not true that the maximal batch size can be approximated as the maximal batch size that fails immediately minus the max_length (which was my original hypothesis) - it can be be much lower. BTW: In https://github.com/tensorflow/tensor2tensor/issues/444#issuecomment-362917578 I report the empirical maximum batch size for a given model, optimizer and max_length.

My guess is that there is something wrong with the [batching and bucketing](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/utils/data_reader.py). I was lazy to check the source code carefully. Maybe it is allowed to exceed the `batch_size` (the first sentence which exceeds the given number of subwords is kept in the batch, although it should be excluded). Maybe the memory is consumed by the queue of sentences for bucketing (although I would say, these are in the CPU RAM, not GPU memory). Anyway, it would be nice if the training **fails either immediately or never**.

**Comment by mehmedes** [comment]
In 1.5.1 I experience OOM when using ```schedule=train``` even after 12h. 
When using ```schedule=continuous_train_and_eval``` training doesn't crash, but actual bleu is lower than with  ```schedule=train```.

**Comment by martinpopel** [comment]
@mehmedes: The `continuous_train_and_eval` bug is reported in #556.
It seems that `continuous_train_and_eval` starts training from a pseudo-random position in the data after each evaluation (according to @rsepassi *it won't see the data as evenly as schedule=train*). So it is possible that `continuous_train_and_eval` will crash as well once it gets to the position where schedule=train crashed, but it cannot be predicted when it will happen (theoretically it can take more then one epoch, considering that `continuous_train_and_eval` samples the training data *with replacement*).

**Comment by mehmedes** [comment]
If I train with ```schedule=train``` and the training crashes after 1 day or so, will T2T restart training from scratch if I restart training or will it continue from where it crashed?

**Comment by martinpopel** [comment]
@mehmedes: the model weights and Adam momentum are stored in the checkpoint, so in this sense the training continues. However, according to https://github.com/tensorflow/tensor2tensor/issues/556#issuecomment-364306094 the position in the training data is not stored in the checkpoint (nor there is any attempt to reconstruct it from the global_step and batch_size), so the resumed training starts from scratch, which means choosing a random training file (but with a fixed rand seed it should be always the same file).

**Comment by mehmedes** [comment]
So, restarting training wouldn't actually help as it would only train upto the same point where it always crashes and not go beyong that?

**Comment by rsepassi** [comment]
Does this continue to happen with `eval_drop_long_sequences=True`? I'm wondering if there's a super big eval batch that gets created and OOMs.

**Comment by martinpopel** [comment]
Oh, I forgot to mention all my experiments with OOM fails reported here were with `--schedule=train`, so `eval_drop_long_sequences` makes no difference because there was no evaluation.

**Comment by martinpopel** [comment]
The OOM is reported as any other OOM - with hundred pages memory dump (if really needed, I can find it in old logs and paste somewhere). It is quite consistent, see e.g. this description: https://github.com/tensorflow/tensor2tensor/issues/637#issuecomment-370750170

**Comment by martinpopel** [comment]
Good idea. That seems easier to implement - all the sentences in the dummy batch should have `max_length` tokens (both source and target side). If no `max_length` is specified than there should be just a single sentence with `batch_size` tokens (which is perhaps too pessimistic, but there is no other way how to prevent OOM; if needed there can be a non-zero default `max_length`).
